src/net-questdb-client/Ingress/Senders/HttpSender.cs

```diff
@@ -0,0 +1,533 @@
+            {
+                throw new IngressError(ErrorCode.ServerFlushError, ex.Message, ex);
+            } 
+            
+            throw;
+        }
+    }
+    
+    /// <summary>
+    ///     Specifies whether a negative <see cref="HttpResponseMessage" /> will lead to a retry or to an exception.
+    /// </summary>
+    /// <param name="code">The <see cref="HttpStatusCode" /></param>
+    /// <returns></returns>
+    // ReSharper disable once IdentifierTypo
+    private static bool IsRetriableError(HttpStatusCode code)
+    {
+        switch (code)
+        {
+            case HttpStatusCode.InternalServerError:
+            case HttpStatusCode.ServiceUnavailable:
+            case HttpStatusCode.GatewayTimeout:
+            case HttpStatusCode.InsufficientStorage:
+            case (HttpStatusCode)509: // Bandwidth Limit Exceeded
+            case (HttpStatusCode)523: // Origin Is Unreachable
+            case (HttpStatusCode)524: // A Timeout Occurred
+            case (HttpStatusCode)529: // Site is overloaded
+            case (HttpStatusCode)599: // Network Timeout Error
+                return true;
+            default:
+                return false;
+        }
+    }
+    
+    /// <inheritdoc cref="FinishOrRetryAsync(System.Net.Http.HttpResponseMessage,System.Threading.CancellationTokenSource?,int)"/>
+    private void FinishOrRetry(HttpResponseMessage response,
+        CancellationTokenSource? cts = default, int retryIntervalMs = 10)
+    {
+        var lastResponse = response;
+
+        if (!response.IsSuccessStatusCode)
+        {
+            if (!(IsRetriableError(response.StatusCode) && Options.retry_timeout > TimeSpan.Zero))
+            {
+                throw new IngressError(ErrorCode.ServerFlushError, response.ReasonPhrase);
+            }
+
+            var retryTimer = new Stopwatch();
+            retryTimer.Start();
+
+            var retryInterval = TimeSpan.FromMilliseconds(retryIntervalMs);
+
+            while (retryTimer.Elapsed < Options.retry_timeout)
+            {
+                var jitter = TimeSpan.FromMilliseconds(Random.Shared.Next(0, retryIntervalMs) - retryIntervalMs / 2.0);
+                Thread.Sleep(retryInterval + jitter);
+
+                var (nextRequest, nextToken) = GenerateRequest();
+                lastResponse = _client!.Send(nextRequest, nextToken!.Token);
+                if (!lastResponse.IsSuccessStatusCode)
+                {
+                    if (IsRetriableError(lastResponse.StatusCode) && Options.retry_timeout > TimeSpan.Zero)
+                    {
+                        cts!.Dispose();
+                    }
+                }
+            }
+        }
+
+        if (!lastResponse.IsSuccessStatusCode)
+        {
+            throw new IngressError(ErrorCode.ServerFlushError, lastResponse.ReasonPhrase);
+        }
+
+        LastFlush = (lastResponse.Headers.Date ?? DateTimeOffset.UtcNow).UtcDateTime;
+        cts!.Dispose();
+        _buffer.Clear();
```

Comment:
Any point leaving dirty buffer on unsuccessful request? Is there a test that tries to re-send the same buffer after an error?